src/lib.rs

```diff
@@ -272,15 +348,22 @@ impl Surface {
     ///
     /// If the caller wishes to synchronize other surface/window changes, such requests must be sent to the
     /// Wayland compositor before calling this function.
-    #[inline]
-    pub fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
-        if (width as usize) * (height as usize) != buffer.len() {
-            panic!("The size of the passed buffer is not the correct size. Its length must be exactly width*height.");
-        }
+    pub fn present(self) -> Result<(), SoftBufferError> {
```

Comment:
It seems like it would be preferable to allow the user to create several buffers ahead of presenting them, to allow batching copies or (if we figure out a less restrictive threading model later) even preparing them on a background thread. For example, the animation example could create a series of Buffer objects, and then submit them reusably to be truly zero copy. To implement this we need present to take buffer as a reference, and we need to have it tied to a non-mut reference to surface. We also likely need some reference counting for the internal data to deal with keeping the buffers alive if the user drops them while the underlying API still needs them. With the Wayland backend at least, this seems possible, but please correct me if this is wrong.

Reply:
Something like that is definitely possible on Wayland. But on other platforms we can't call any/most window management related functions from off the main thread. So in a multi-threaded context even implementing `Drop` may be problematic if a `Buffer` can be used on another thread; maybe if it sets a flag and depends on one of these functions called in the main thread to eventually clean up resources.

If this is allowed only in a single thread I don't know if there are particular difficulties with it. Though it's non-trivial to support across backends.

Reply:
In any case, this is a big change that is probably better left for a later pull request. It would probably be a breaking change, but at least until the next release we can break the API as many times as we want.